router/batchrouter/asyncdestinationmanager/eloqua/utils.go

```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		return "", err
+	}
+	csvWriter.Flush()
+	var line string
+	index := 0
+	for scanner.Scan() {
+		line = scanner.Text()
+		var data TransformedData
+		if err := json.Unmarshal([]byte(line), &data); err != nil {
+			return "", err
+		}
+		var values []string
+		for _, field := range fields {
+			values = append(values, data.Message.Data[field].(string))
+		}
+		fileInfo, err := csvFile.Stat()
+		if err != nil {
+			return "", err
+		}
+		if fileInfo.Size() > uploadJobInfo.fileSizeLimit {
+			left, _ := lo.Difference(uploadJobInfo.importingJobs, uploadJobInfo.succeededJobs)
+			uploadJobInfo.failedJobs = append(uploadJobInfo.failedJobs, left...)
+			break
+		}
+		jobIdRowMap[int64(index+1)] = uploadJobInfo.importingJobs[index]
+		uploadJobInfo.succeededJobs = append(uploadJobInfo.succeededJobs, uploadJobInfo.importingJobs[index])
+		index += 1
+		err = csvWriter.Write(values)
+		if err != nil {
+			return "", err
+		}
+		csvWriter.Flush()
+	}
+	return csvFilePath, nil
+}
+
+func createBodyForImportDefinition(evenType string, fields []string, eloquaFields *Fields, identifierFieldName string) (map[string]interface{}, error) {
+	fieldStatement := make(map[string]string)
+	for _, val := range fields {
+		for _, val1 := range eloquaFields.Items {
+			if val1.InternalName == val {
+				fieldStatement[val] = val1.Statement
+			}
+		}
+	}
+	if evenType == "identify" {
+		return map[string]interface{}{
+			"name":                    "Rudderstack-Contact-Import",
+			"fields":                  fieldStatement,
+			"identifierFieldName":     identifierFieldName,
+			"isSyncTriggeredOnImport": false,
+		}, nil
+	} else if evenType == "track" {
+		return map[string]interface{}{
+			"name":                    "Rudderstack-CustomObject-Import",
+			"updateRule":              "always",
+			"fields":                  fieldStatement,
+			"identifierFieldName":     identifierFieldName,
+			"isSyncTriggeredOnImport": false,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unable to create body for import definition")
+}
+
+func parseRejectedData(data *HttpRequestData, importingList []*jobsdb.JobT, service Eloqua, jobToCSVMap map[int64]int64) (*common.EventStatMeta, error) {
+	jobIDs := []int64{}
+	for _, job := range importingList {
+		jobIDs = append(jobIDs, job.JobID)
+	}
+	rejectResponse, err := service.CheckRejectedData(data)
+	if err != nil {
+		return nil, err
+	}
+	failedJobIDs := []int64{}
+	failedReasons := map[int64]string{}
+	/**
+	We have a limit of 1000 rejected events per api call. And the offset starts from 0. So we are fetching 1000 rejected events every time making the api call and updating the offset accordingly.
+	*/
+	if rejectResponse.Count > 0 {
+		iterations := int(math.Ceil(float64(rejectResponse.TotalResults) / float64(1000)))
+		var offset int
+		for i := 0; i < iterations; i++ {
+			offset = i * 1000
+			data.Offset = offset
+			if offset != 0 {
+				rejectResponse, err = service.CheckRejectedData(data)
```

Comment:
do we need to retry this?

Reply:
We are already retrying if we are getting any error here. We are returning 500 status from `GetUploadStats` step in case of any error.